### PR TITLE
Smooth scrolling and more (pixel-based tile draw)

### PR DESCRIPTION
**Pull request by crudelios**
This PR implements smooth scrolling, fundamentally changing how scrolling works.

It does more, however:

* It changes tile image clipping to be pixel-based instead of half-tile based. On the checks I ran, it does not seem to incur a significant performance penalty (tested on my Surface Pro 4 at a resolution of 2736 x 1824.
* It allows drawing of more tiles to prevent visual glitches at the bottom of the screen (especially noticeable with hippodromes). This causes a slight (about 2 to 4%) performance penalty, but it can be reduced or reverted.
* This build removes the right border and bottom border. The bottom border was useless, but the right border had a pattern that some may like. Still, there is now an extra bit of real estate.
* Make the scroll speed framerate-independent. Before the scroll was slower at low framerates, now the game dynamically adjusts the scroll to compensate, leading to consistent scroll speeds.
* Add some modern decays to scrolling. When you stop scrolling, the city will roll for a few more pixels before stopping.
* Improve touch scroll when building by reducing margins and reworking the speed.
* And probably some extra things that I don't recall.

This needs some testing on slower devices to make sure my performance results are good.

### CODE REVIEW

**Comment by Legionary2** [comment]
Hey there crudelios, I know this is unrelated but I remember watching GamerZach showcase this open source clone last year (not the 2 hour video recently) and someone on the comment section mentioned they forked the code and implemented a less ugly gatehouse for personal use. The user name looked familiar to yours although my memory may be playing tricks on me. I cant seem to find that comment any longer either. By any chance was that you?

**Comment by bvschaik** [comment]
Can we make the smooth scrolling a (runtime) option? I understand that this is desirable for Android and other touch-based devices, but I quite like the tile-by-tile scrolling...
I plan to implement a configuration screen for #284, so this would be a good candidate to put on there.

**Comment by crudelios** [comment]
Sure, it shouldn't be too hard to emulate the old mechanism with the new implementation.

**Comment by crudelios** [comment]
Hi @Legionary2, I did comment on that video but I don't remember saying anything about gatehouses and I didn't add them. But then again I don't remember anyone else talking about forking the code back then, so take this with a grain of salt.

File: src/widget/minimap.c [hunk 1]
```diff
@@ -199,11 +199,13 @@ static void draw_minimap_tile(int x_view, int y_view, int grid_offset)
...
 {
     int camera_x, camera_y;
+    int camera_pixels_x, camera_pixels_y;
     city_view_get_camera(&camera_x, &camera_y);
+    city_view_get_pixel_offset(&camera_pixels_x, &camera_pixels_y);
     int view_width_tiles, view_height_tiles;
     city_view_get_viewport_size_tiles(&view_width_tiles, &view_height_tiles);
 
-    int x_offset = data.x_offset + 2 * (camera_x - data.absolute_x) - 2;
+    int x_offset = data.x_offset + 2 * (camera_x - data.absolute_x) - 2 + camera_pixels_x / 30;
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/widget/minimap.c]: `camera_x` is originally doubled, so the minimap viewport rect only increases by two px. This makes it increase by pixel, making it look smoother when scrolling.

File: src/platform/keyboard_input.c [hunk 2]
```diff
@@ -99,47 +99,19 @@ void platform_handle_key_down(SDL_KeyboardEvent *event)
...
             keyboard_left();
-            hotkey_left();
-            // check if another key still down as well (only last-pressed repeats)
-            if (event->repeat && is_key_currently_down(SDL_SCANCODE_UP)) {
-                hotkey_up();
-            }
-            if (event->repeat && is_key_currently_down(SDL_SCANCODE_DOWN)) {
-                hotkey_down();
-            }
+            hotkey_left_press();
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/platform/keyboard_input.c]: It should work with two arrow keys, `left` has priority over `right` and `top` over `bottom`. Diagonals work.

> how does this behave when I press an arrow key, switch to another window, release the key, and then return? The old Impressions games had trouble with that.

No idea, I'll test!

File: src/core/image.h [hunk 3]
```diff
@@ -8,6 +8,11 @@
 #define IMAGE_FONT_MULTIBYTE_OFFSET 10000
 #define IMAGE_FONT_MULTIBYTE_MAX_CHARS 2188
 
+typedef enum {
+    IMAGE_TYPE_NORMAL = 0,
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/core/image.h]: I tried looking for other referenced image types but didn't find any, so I only added those two as I don't really know what the other types are.

**Comment by crudelios** [comment]
> General observations while testing with the setting OFF:
> 
> * Empire map scroll is not affected by scroll speed setting
> * Scrolling with arrow keys on the city map:
>   
>   * should NOT speed up with scroll speed: one press == one tile
>   * is very sensitive: it often scrolls 2 tiles instead of 1 while I don't press the key any longer than usual

This should all be fixed.

File: src/widget/city.c [hunk 4]
```diff
@@ -194,33 +196,25 @@ static void widget_city_handle_touch_scroll(const touch *t)
...
-        return;
-    }
-    if (!city_view_pixels_to_view_tile(t->current_point.x, t->current_point.y, &current)) {
-        return;
-    }
-
-    if (scroll_move_touch_drag(original.x, original.y, current.x, current.y, &camera_position)) {
-        city_view_set_camera(camera_position.x, camera_position.y);
+    touch_coords original = scroll_get_original_touch_position();
+    if(scroll_move_touch_drag(original.x, original.y, t->current_point.x, t->current_point.y, &camera_pixel_position)) {
```
Comment (bvschaik) [inline src/widget/city.c]: Space after `if`.
```suggestion
    if (scroll_move_touch_drag(original.x, original.y, t->current_point.x, t->current_point.y, &camera_pixel_position)) {
```

**Comment by QmwJlHuSg9pa** [comment]
Smooth scrolling: Please add an option to disable the decay/deceleration. I find it really off-putting.

**Pull request merged**